specs/commandline/policy.md

```diff
@@ -0,0 +1,232 @@
+- Policy "allow-expired-images" is for logging instead of failing expired images stored in repository `registry.acme-rockets.io/software/legacy/metrics`.
+- Policy "global-policy-for-all-other-images" is for verifying any other images that signed by the ACME Rockets.
+
+```jsonc
+{
+    "version": "1.0",
+    "trustPolicies": [
+        {
+            "name": "wabbit-networks-images",
+            "registryScopes": [ 
+              "registry.acme-rockets.io/software/net-monitor",
+              "registry.acme-rockets.io/software/net-logger" 
+            ],
+            "signatureVerification": {
+                "level": "strict"
+            },
+            "trustStores": [ 
+              "ca:wabbit-networks",
+            ],
+            "trustedIdentities": [
+                "x509.subject: C=US, ST=WA, L=Seattle, O=wabbit-networks.io, OU=Security Tools"
+            ]
+        },
+        {
+            "name": "unsigned-image",
+            "registryScopes": [ "registry.acme-rockets.io/software/unsigned/net-utils" ],
+            "signatureVerification": {
+              "level" : "skip" 
+            }
+        },
+        {
+            "name": "allow-expired-images",
+            "registryScopes": [ "registry.acme-rockets.io/software/legacy/metrics" ],
+            "signatureVerification": {
+              "level" : "strict",
+              "override" : {
+                "expiry" : "log"
+              }
+            },
+            "trustStores": ["ca:acme-rockets"],
+            "trustedIdentities": ["*"]
+        },
+        {
+            "name": "global-policy-for-all-other-images",
+            "registryScopes": [ "*" ],       
+            "signatureVerification": {                                
+                "level": "strict"
+            },
+            "trustStores": [ 
+              "ca:acme-rockets"
+            ],                  
+            "trustedIdentities": [                                    
+                "x509.subject: C=US, ST=WA, L=Seattle, O=acme-rockets.io, CN=SecureBuilder"
+            ]
+        }
+    ]
+}
+```
+
+## Outline
+
+### notation policy command
+
+```text
+Manage trust policies for signature verification.
+
+Usage:
+  notation policy [command]
+
+Available Commands:
+  import    import trust policies from a JSON file
+  init      initialize trust policies with default values
+  show      show trust policies
+
+Flags:
+  -h, --help   help for policy
+```
+
+### notation policy import
+
+```text
+Import trust policies from a JSON file
+
+Usage:
+  notation policy import [flags] <file_path>
+
+Flags:
+  -d, --debug     debug mode
+      --force     override the existing trust policies, never prompt
+  -h, --help      help for import
+  -v, --verbose   verbose mode
+```
+
+### notation policy init
+
+```text
+Initialize trust policies with default values
+
+Usage:
+  notation policy init [flags]
```

Comment:
The issue with init command is that users will just use it without knowing its implication. 
In init command we are making following assumptions
*  `"registryScopes": ["*"]`  Looks fine 
*  `"signatureVerification": {"level": "strict"}`  If user are not aware of this then there verification might start failing in production because of revoked/expired signature. IMO we shouldn't default here.
*  `"trustStores": ["ca:default"]` In most of the usecases this value will be different as user will create there own trust-store. IMO we shouldn't default here.
*  `"trustedIdentities": ["*"]`  <-- If users are using publicly trusted certificates then it gives false sense of security. IMO we shouldn't default here.

Given that we can safely default only for 1 option out of 3, IMO we shouldn't include init option.

Reply:
@shizhMSFT @patrickzheng200 do you have any comments on Pritesh's comments? There are some security concerns here. One way to improve it is to introduce more flags for users to specify the value for each property, but it is the `notation policy add` command we want to do initially, but considering the scope and the time of release, it's better to do it in an iterate way. Let's agree on the way forward and move on.

Reply:
@priteshbandi The purpose of the `notation policy init` is to improve the overall out of the box experience (OOBE) of the `notation` CLI. If an advanced customer does not think it is secure, that customer can compose their own policy file and use `notation policy import` to import it.

Reply:
Again, writing a valid trust policy is not an easy job for a new user. They need to learn from scratch.

Reply:
For a 0-experience starter, find our specs -> find a trust policy template -> learn how to use it -> change it to user wanted values could be a very hard workflow. The learning cost for a beginner is a bit too high here. The concern is that some new users just want to give it a try, but found it's hard to use, then they might just give up at this point.

`notation policy init` serves as an entry point for notation's trust policy configuration. We should print out the generated default trust policy statement for users to evaluate if it's good enough. For users who don't like this policy or think this policy is too naive, then they can use the default policy as the template (so they don't need to do the search on their own), then use `notation policy import` command to overwrite the default.

Reply:
Here is my proposal, please be aware that the `notation policy` command is new and marked as `preview` for RC-4 release. We can collect feedback from users and continuously improve it. 
iteration 1: introduce `notation policy import` and `notation policy show` commands （RC-4 and 1.0 experience)
iteration 2: introduce `notation policy init` for OOBE experience (post 1.0)
iteration 3: introduce `notation policy` CRUD commands for fine tuning (post 1.0)

@priteshbandi @patrickzheng200 @shizhMSFT Could you vote? Then I can start updating the spec.

Reply:
I'd suggest including `notation policy init` in `1.0`.

Reply:
@shizhMSFT Let's get `iteration 1` done in RC-4, and in parallel, create a new issue to discuss the experience for `notation policy init`. I will create one for tracking.

Reply:
> Here is the proposal, please be aware that the `notation policy` command is new and marked as `preview` for RC-4 release. We can collect feedback from users and continuously improve it. iteration 1: introduce `notation policy import` and `notation policy show` commands （RC-4 and 1.0 experience) iteration 2: introduce `notation policy init` for OOBE experience (post 1.0) iteration 3: introduce `notation policy` CRUD commands for fine tuning (post 1.0)
> 
> @priteshbandi @patrickzheng200 @shizhMSFT Could you vote? Then I can start updating the spec.

+1 for having `notaiton policy init` in `1.0`.

Reply:
> @priteshbandi The purpose of the notation policy init is to improve the overall out of the box experience (OOBE) of the notation CLI. If an advanced customer does not think it is secure, that customer can compose their own policy file and use notation policy import to import it.

How are we going to communicate that the trust policy can be insecure and unsage? As an end user I would assume the out of box defaults for a security products should be safe and secure.

> For a 0-experience starter, find our specs -> find a trust policy template -> learn how to use it -> change it to user wanted values could be a very hard workflow. The learning cost for a beginner is a bit too high here. The concern is that some new users just want to give it a try, but found it's hard to use, then they might just give up at this point.
notation policy init serves as an entry point for notation's trust policy configuration. We should print out the generated default trust policy statement for users to evaluate if it's good enough. For users who don't like this policy or think it's too basic, then they can use the default policy as the template (so they don't need to do the search on their own), then use notation policy import command to overwrite the default.

How will we communicate to user that they need to look beyond init? I am worried that if something things works, user might not look into its implications. Should we have command `notation policy generate-sample` which will print out policy, if user wants they can pipe it to notation import ?

I do recognize the need of out of the box experience but being security product we should be cautious to not provide insecure and unsafe defaults.

Reply:
@priteshbandi @shizhMSFT @patrickzheng200 Thanks for the great discussions. I suggest we implement `import` and `show` commands in RC-4, and continue the discussion of OOBE tracked at issue #590.

Reply:
> * ost of the usecases this value will be different as user will create there own trust-store. IMO we shouldn't default here.

+1. I think @priteshbandi is right here, do we have customer data to support the defaults ?